File: backend/app/models/inventory.py
"""Host Inventory models (Mission M003.5 §3).

Host Inventory is *information about THIS machine* — hardware identity, OS
identity, structured storage inventory, network interfaces, and maintenance
status — observed by the host collector and reported as ``host_inventory``
events. It is deliberately distinct from the Fleet Registry (*information
aggregated about ALL nodes*): the registry holds administered identity;
inventory is observed fact and may change with every boot or hardware swap.

The payload is intentionally schema-flexible (a validated-but-open dict):
sections are collector-versioned and designed to gain keys — SMART data per
storage device, new hardware families, site metadata — without backend
schema changes (the M003.5 single-node → multi-site scale-out requirement).
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict

from app.models.registry import FleetId

#: Event type carrying host inventory reports.
HOST_INVENTORY_EVENT_TYPE = "host_inventory"


class HostInventoryRecord(BaseModel):
    """Latest host inventory projection for one node.

    ``payload`` sections (all optional, collector fail-soft):

    * ``hardware`` — manufacturer, model, revision, cpu_model,
      cpu_architecture, cpu_cores, memory_total_bytes, serial;
    * ``os`` — name, release, version_id, pretty_name, kernel, hostname;
    * ``storage`` — list of device dicts (name, device, type, transport,
      capacity_bytes, mount, brand, filesystem, usage fields; extra keys
      such as future SMART attributes are preserved verbatim);
    * ``network`` — interfaces (name, ipv4, link_state) + default_route;
    * ``maintenance`` — last_apt_update_epoch, last_apt_upgrade,
      last_apt_full_upgrade, updates_available, reboot_required.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    fleet_id: FleetId
    payload: dict[str, Any]
    #: Source timestamp of the inventory event (collector clock).
    reported_at: datetime
    #: When the backend projected this revision.
    updated_at: datetime

File: frontend/src/types/inventory.ts
/**
 * Mirrors backend/app/models/inventory.py (Host Inventory, Mission M003.5 §3).
 *
 * The backend stores the payload as a validated-but-open dict — sections are
 * collector-versioned and may gain keys without a backend schema change. The
 * mirror therefore types every section and field as optional/nullable: the
 * UI must render gracefully when a section is missing or partial.
 */

/** hardware section — manufacturer/model/CPU/memory identity (§3a). */
export interface HardwareInventory {
  manufacturer?: string | null;
  model?: string | null;
  revision?: string | null;
  cpu_model?: string | null;
  cpu_architecture?: string | null;
  cpu_cores?: number | null;
  memory_total_bytes?: number | null;
  serial?: string | null;
}

/** os section — /etc/os-release identity + kernel + hostname (§3c). */
export interface OsInventory {
  name?: string | null;
  release?: string | null;
  version_id?: string | null;
  pretty_name?: string | null;
  kernel?: string | null;
  hostname?: string | null;
}

/**
 * One entry per mounted, block-device-backed filesystem (§3b). Extra keys
 * (e.g. future SMART attributes) are preserved by the backend verbatim.
 */
export interface StorageDevice {
  name?: string | null;
  device?: string | null;
  physical_device?: string | null;
  type?: string | null;
  transport?: string | null;
  capacity_bytes?: number | null;
  mount?: string | null;
  brand?: string | null;
  filesystem?: string | null;
  total_bytes?: number | null;
  used_bytes?: number | null;
  free_bytes?: number | null;
  used_percent?: number | null;
}

export interface NetworkInterface {
  name?: string | null;
  ipv4?: string | null;
  link_state?: string | null;
}

export interface DefaultRoute {
  gateway?: string | null;
  interface?: string | null;
}

export interface NetworkInventory {
  interfaces?: NetworkInterface[] | null;
  default_route?: DefaultRoute | null;
}

/** maintenance section — apt/reboot status (§3d). */
export interface MaintenanceInventory {
  last_apt_update_epoch?: number | null;
  last_apt_upgrade?: string | null;
  last_apt_full_upgrade?: string | null;
  updates_available?: number | null;
  reboot_required?: boolean | null;
}

/** The schema-flexible payload; every section is optional (fail-soft collector). */
export interface HostInventoryPayload {
  hardware?: HardwareInventory | null;
  os?: OsInventory | null;
  storage?: StorageDevice[] | null;
  network?: NetworkInventory | null;
  maintenance?: MaintenanceInventory | null;
}

/** Read model of GET /api/v1/fleet/{fleetId}/inventory (404 when never reported). */
export interface HostInventoryRecord {
  fleet_id: string;
  payload: HostInventoryPayload;
  reported_at: string;
  updated_at: string;
}
